### ISSUE TITLE
Recovery and rebalance throughput are controlled with the same parameter

### ISSUE THREAD

**Issue by prog8** [elastic/elasticsearch#20716]
ES currently have a single parameter which controls the speed of recovery and rebalance. Recovery in most cases has higher priority than cluster rebalancing. When the cluster has problems or node is restarted one may want the node to be quickly recovered even paying the price of slower indexing for short period of time. This is not always true for rebalancing.

I also have an impression that `recovery.max_bytes_per_sec` is used not only for network traffic but also disk reads. Why I think so? I had setting `recovery.max_bytes_per_sec` 60mb and decided to do rolling restart. Before each node restart I triggered flush synced (BTW most of shards  are immutable). It was taking very long until node was recovered even thought there was very little network traffic (most of data was just read from disk). When I changed `recovery.max_bytes_per_sec` to 600mb recovery was much much faster. Look at the chart below. Is this true that `recovery.max_bytes_per_sec` applies also for disk reads? Or maybe there is another reasons that changing `recovery.max_bytes_per_sec` to 600mb helped so much.
This is observation is the most important reason why I asked whether recover and rebalance should share the same setting for `max_bytes_per_sec`.

The chart below is from ES 2.3 cluster. 

![selection_657](https://cloud.githubusercontent.com/assets/6692291/18992371/33ea0018-871e-11e6-8889-75a3c7e569fc.png)

**Comment by bleskes** [comment]
A couple of clarifications:
1) By default a cluster won't rebalance until it's green (although it might have ongoing rebalance recoveries when a node left which are not completed by the time the node joins). see [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/shards-allocation.html#_shard_rebalancing_settings)
2) The `recovery. max_bytes_per_sec` settings is used to throttle recovery bytes when sent across the network, not locally. 

All of this means that I'm not sure why you're recoveries got faster after the changed of the settings, *if* all shards were sync flushed. Did you check the output of your sync flush command? were all shards synced as expected?

**Comment by prog8** [comment]
> 1) By default a cluster won't rebalance until it's green (although it might have ongoing rebalance recoveries when a node left which are not completed by the time the node joins). see here
Yes I know but this does not change the fact that I have to manually adjust the value of `recovery.max_bytes_per_sec` after recovery is done (not to let rebalance take to much IOPS and network)

> 2) The recovery. max_bytes_per_sec settings is used to throttle recovery bytes when sent across the network, not locally.
This is what I thought before I changed `recovery.max_bytes_per_sec` which significantly speeded up the recovery.

I didn't look at the output of flush because it was too big, ... but let's look at another screenshot. Doesn't it look suspicious? You can see that when `recovery.max_bytes_per_sec` has lower value there is longer time of "no network activity" and recovery is still in progress.
First chart presents the number of unassigned  shards and the second shows the total network traffic in the cluster. If shards are not synced correctly we'd see much higher network traffic, no?

![selection_662](https://cloud.githubusercontent.com/assets/6692291/19036184/750cf652-896e-11e6-9888-f493215d27e1.png)

**Comment by bleskes** [comment]
This is indeed interesting. If the synced flush did it's job (it should have small header at the beginning of the response  to give count totals), no lunce indices should be copied (assuming no indexing). Also not that the height of the first wave of bars is not really different than the height of the last cycle. I suspect something else is in play... can you tell more about your setup? number of indices /shards per node?

**Comment by prog8** [comment]
I have about 8K shards in the cluster. Each node holds about 1K shards. Many of shards are tiny (<100mb), others are up to 5GB size. The ratio of tiny/non-tiny shards is more or less 50% to 50%. In the moment when I took screenshots I was replacing a few ES instances with another ones (different hardware), thus the relocation. In meantime I realized I have wrong Xmx settings so I had to do rolling restart of 4 new nodes (all these nodes are identical in terms of CPU, storage, RAM, network). I flushed indices restarted the first node, flushed indices again and restarted the other one, ..., finally I changed `recovery.max_bytes_per_sec`, flushed and restarted the last node. The output is as you can see. The cluster has constant indexing rate and constant query rate. Ah I found one thing. Master node reached OOM but when you look at timing it was after the cluster was recovered (recovery was done at 12:10, OOM is at 12:20).

![selection_666](https://cloud.githubusercontent.com/assets/6692291/19037210/5a31a330-8975-11e6-873d-9312efc66cfc.png)

**Comment by bleskes** [comment]
> The cluster has constant indexing rate and constant query rate

Do you keep indexing while restarting the nodes?

Do you have your master log in debug mode by any chance? if so it would be great if you can share them (mail them to my first name at elastic dot co).

**Comment by prog8** [comment]
> Do you keep indexing while restarting the nodes?
Yes, this is the case. I need NRT and I cannot stop indexing while restarting nodes. I know this slows down rolling restart but I have no other option. I would not be suspicious that restart takes so long if I don't see speed up after changing `recovery.max_bytes_per_sec`. This is the reasons why I thought it might make sense to have something like `recovery.max_bytes_per_sec` and `rebalance.max_bytes_per_sec` and they could have different values.

Unfortunately I don't have debug logs :(  But maybe this is a good practice to enable debug logging on master nodes before rolling restart. I'll keep that in mind.

**Comment by bleskes** [comment]
Thx. Do you keep indexing while restarting the nodes?

**Comment by prog8** [comment]
Yes I keep indexing. As I mentioned in previous message.
> I need NRT and I cannot stop indexing while restarting nodes. I know this slows down rolling restart but I have no other option. I would not be suspicious that restart takes so long if I don't see speed up after changing recovery.max_bytes_per_sec

**Comment by bleskes** [comment]
Sorry, missed your first sentence. 

So I think we can say that you recovery isn't limited by throttling (btw you can check that by calling `GET _recovery?human` and check throttle time). What kind of hardware are you running on? Is it any kind of remote storage?

**Comment by prog8** [comment]
I'm using EC2 instances 4th generation with EBS attached. So ... yes, this is remote storage. I think if EBS is a problem I'd see high io wait times which is not the case but maybe I'm missing something.

> So I think we can say that you recovery isn't limited by throttling

Do you think this is  just a coincidence that changing `recovery.max_bytes_per_sec` did the job?

> GET _recovery?human

I'll look at this call on next restart. I was always using `_cat/recovery?active_only`

**Comment by bleskes** [comment]
> I'm using EC2 instances 4th generation with EBS attached  ... EBS is a problem I'd see high io wait times which is not the case but maybe I'm missing something

The thing is that you don't see more bandwidth usage in the last faster recovery. I *suspect* the time is cause by the master reaching out to nodes and looking for data. We currently do it on demands, i.e., when we can actually can allocated a new shard (we have throttling on the number of concurrent recoveries a node can take - see [here]( https://www.elastic.co/guide/en/elasticsearch/reference/current/shards-allocation.html#_shard_allocation_settings) . This means that once recovery is done and more shards can be allocated, the master will reach out again. These small IOs can be slow on remote storage. I have some ideas on how to potentially do it oncurrently with recovery. But we'll see.

> I'll look at this call on next restart. I was always using _cat/recovery?active_only

You can do it now, without the active only flag. It will return the information about the last recoveries.

**Comment by prog8** [comment]
Thank you very much Boaz. I really appreciate your effort. Do you think then that recovery speed up was just a coincidence?

> I suspect the time is cause by the master reaching out to nodes and looking for data. We currently do it on demands, i.e., when we can actually can allocated a new shard

What do you mean by this? Maybe this is indeed disk access latency. Can you tell me more details what kind of small IOs are done? I'm really curious. Maybe this indeed can be handled only by significantly changing number of concurrent recoveries (current settings is: 5).

> You can do it now, without the active only flag. It will return the information about the last recoveries.

Yeah I'll analyze recovery history but it may take some time :). I wanted to avoid it because recovery is not marked with timestamp and I have almost 8000 entries in recovery history

**Comment by bleskes** [comment]
> Do you think then that recovery speed up was just a coincidence?

Feels like it.

> Can you tell me more details what kind of small IOs are done?

The master reaches out to all nodes asking for information about local data. The nodes then look at their local disk and see what data they have. The also open the lucene index (just read a couple of files) and double check versions and corruption markers. 

> Yeah I'll analyze recovery history but it may take some time :). I wanted to avoid it because recovery is not marked with timestamp and I have almost 8000 entries in recovery history

Grep `throttle_time` is your friend :)

**Comment by prog8** [comment]
> Grep throttle_time is your friend :)

Hehe ... but you know ... it will not tell the full truth. I can see throttling for some operations while no throtthgling for others (including rebalance which is also mentioned in recovery log). This is indeed what I can see. Some recoveries have throttling time 0 and others 294163 or 405536

**Comment by prog8** [comment]
I did an experiment. I restarted node (flush, 5 concurrent recoveries) and it took 20 minutes. Later I changed concurrent recoveries to 30 and again flushed before restart. Node recovered in 6 minutes. In meantime i didn't notice high IO utilization. I think as you mentioned @bleskes this might be caused by slow IO access time. I can imagine that e.g. access time is lets say 100ms (don't take this value seriosly, I just gave an example) and with 5 concurrent recoveries we are not able to utilize whole throughput and IOPS, so after changing to 30 concurrent recoveries and the same disk access time we can utilize more IOPS and disk throughput and recovery is just faster. Makes sense?

**Comment by bleskes** [comment]
> Makes sense?

Yep. I think we can consider fetching shard data concurrently. If you want, I can guide you through a PR.

> Of course this does not mean I don't see reason to use both recovery.max_bytes_per_sec and rebalance.max_bytes_per_sec which might be still useful (at least I think so).

In theory I agree, but the system is already complex enough (see our discussion :)) - so I don't think we should add any settings unless there is really a concrete and repeated pain.

**Comment by prog8** [comment]
I did another experiment. I increased the number of concurrent recoveries to 60 which made disk more utilized (checked utilization with iostat and dstat). Utilization was between 70 and 100% all the time and recovery took 3 minutes. It seems remote storage's higher access time really matters for recovery.

@bleskes thanks a lot  for explaining it and for your time :)

> In theory I agree, but the system is already complex enough (see our discussion :)) - so I don't think we should add any settings unless there is really a concrete and repeated pain.

Yep, I agree, all new parameters add complexity to Elasticsearch. What is more I am afraid  adding something like `rebalance.max_bytes_per_sec` would require much refactoring (didn't look at code but intuition says rebalance is just a specific type of a recovery).

**Comment by DaveCTurner** [comment]
@bleskes it seems that this issue remains open just because of your suggestion that we could fetch the shard data concurrently. If so, I think we should close this thread and (optionally) open a separate issue asking for just that, so future readers don't have to parse that nugget out of this full thread. WDYT?

**Comment by bleskes** [comment]
@DaveCTurner I tend to just close this. I haven't heard anything about these issues for a long while and ops based recovery should also help with recovery times. The system is complex enough (even too much) and I'm worried about adding another asynchronicity

**Comment by DaveCTurner** [comment]
Sounds good to me.
